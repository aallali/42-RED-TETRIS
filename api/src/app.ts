
import * as express from "express";
import * as cors from "cors"
import { Socket } from "socket.io";
import Game from "./Game"

import ROOM from "./Room"
import PLAYER from "./Player"
import { ROOM_MODE } from "./types"


const log = console.log
interface IPlayer {
	id: string;
	nickname: string;
}
interface IRoom {
	id: string;
	name: string;
	players: []
	admin: IPlayer
}
class App {

	public app: express.Application;
	public http: any
	public io: Socket
	private port: number = 4242
	private gameIns: Game
	constructor() {
		this.app = express();
		this.app.use(
			cors({
				origin: "*",
				credentials: true,
			})
		);
		this.http = require("http").Server(this.app);
		this.io = require("socket.io")(this.http, {
			pingInterval: 60000,
			cors: {
				origin: "*",
			},
		});
		this.routes()
		this.initSocketEvents()
		this.gameIns = new Game(this.io)

	}
	/**
	 * 
	 */
	private routes() {
		this.app.get("/", (req: any, res: any) => {
			console.log(Array.from(this.gameIns.players))
			res.send(Array.from(this.gameIns.players))
		});
	}
	private initSocketEvents() {
	}
	public listen() {
		console.clear()
		this.http.listen(this.port, () =>
			console.log(`ðŸš€ Server is running on PORT ${this.port}`)
		);
	}
}

export default new App();
