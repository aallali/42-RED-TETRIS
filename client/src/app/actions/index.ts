import {
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER, UPDATE_SCORE,
	SET_PLAYER_ADMIN, ADD_ROW,
	PLAYER_LOST, RESET_STATES,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN
} from "../../reducers/player.reducer";
import { UPDATE_PLAYERS, CLEAR_OPPONENTS } from "../../reducers/opponent.reducer"
import { SET_ERROR, CLEAR_ERROR } from "../../reducers/error.reducer"

import {
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
} from "../../reducers/game.reducer"

export {
	// Players actions
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER,
	UPDATE_SCORE,
	SET_PLAYER_ADMIN,
	ADD_ROW, PLAYER_LOST,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN,
	RESET_STATES,

	// Opponents actions
	UPDATE_PLAYERS,
	CLEAR_OPPONENTS,
	// Error actions
	SET_ERROR, CLEAR_ERROR,

	// game actions
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
}
