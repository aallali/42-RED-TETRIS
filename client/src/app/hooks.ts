import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { useGameStatus } from "./TetrisHooks/useGameStatus"
import { useInterval } from "./TetrisHooks/useInterval"
import { usePlayer } from "./TetrisHooks/usePlayer"
import { useStage } from "./TetrisHooks/useStage"


import io from "socket.io-client";

export const socket = io("http://localhost:4242");
export { useGameStatus, usePlayer, useInterval, useStage }
// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
