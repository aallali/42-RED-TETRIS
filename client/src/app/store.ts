import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import playerReducer from "../reducers/player.reducer";
import stageReducer from "../reducers/stage.reducer"
export const store = configureStore({
  reducer: {
    player: playerReducer,
    stage: stageReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
