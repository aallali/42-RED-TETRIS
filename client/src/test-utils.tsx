// test-utils.jsx
import React, { ReactChildren } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import playerReducer from "./reducers/player.reducer";
import opponentReducer from "./reducers/opponent.reducer"
import errorReducer from "./reducers/error.reducer"
import gameReducer from "./reducers/game.reducer"
import chatReducer from './reducers/chat.reducer';

function aallaliStore(preloadedState: any) {
	return configureStore({
		reducer: {
			player: playerReducer,
			opponents: opponentReducer,
			error: errorReducer,
			game: gameReducer,
			chat: chatReducer
		},
		preloadedState
	})
}
function render(
	ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
	{
		preloadedState,
		store = configureStore({
			reducer: {
				player: playerReducer,
				opponents: opponentReducer,
				error: errorReducer,
				game: gameReducer,
				chat: chatReducer
			},
			preloadedState
		}),
		...renderOptions
	}: any = {}
) {
	function Wrapper({ children }: { children: ReactChildren }) {
		return <Provider store={store}>{children}</Provider>

	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}


// re-export everything
export * from '@testing-library/react'
// override render method
export { render, aallaliStore }