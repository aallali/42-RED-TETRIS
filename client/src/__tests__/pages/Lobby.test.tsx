
import { Provider } from 'react-redux';
import { store } from '../../app/store';
import Lobby from '../../pages/Lobby';
import renderer from "react-test-renderer";
import * as reactRedux from "react-redux";
// import { render, screen } from "@testing-library/react";
import { render, screen } from '../../test-utils'

import { fireEvent, waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import { rest } from "msw";
import { setupServer } from "msw/node";
import { mount } from 'enzyme';

// import { act } from 'react-test-renderer';
// import { act } from '@testing-library/react';

// import { act } from 'react-dom/test-utils';
const axios = require('axios');

const server = setupServer(
	rest.get("http://localhost:4242/rooms", (_, res, ctx) => {
		return res(ctx.status(200), ctx.json([{ title: "arena", size: 7, active_players: 5, mode: "multiplayer" }]))
	}),
	rest.get("http://localhost:4242/top", (_, res, ctx) => {
		return res(ctx.status(200), ctx.json(`allali,67\nbot2,30\nbotUsername3,27`))
	})
);

// jest.mock('axios')

const testRooms = [{ title: "arena", size: 7, active_players: 5, mode: "multiplayer" }, { title: "arena2", size: 3, active_players: 1, mode: "multiplayer" }]

describe("_Loby PageXContainer test", () => {
	const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");
	beforeEach(() => {
		useDispatchMock.mockClear();
		useDispatchMock.mockReturnValue(jest.fn());
	});
	beforeAll(() => server.listen());

	afterEach(() => server.resetHandlers());

	afterAll(() => server.close());

	test("_ Loby Snapshot Check", () => {
		const tree = renderer.create(<Provider store={store}>
			< Lobby />
		</Provider>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	// test("_ Loby rendering with API mocked values for two sub component (Rooms/TopRankedPlayers) check", async () => {




	// axios.get.mockResolvedValue({
	// 	data: testRooms
	// });
	// render(
	// 	< Lobby />
	// )
	// await waitFor(() => {
	// 	expect(screen.queryByText(/Red-Tetris/i)).toBeInTheDocument()
	// });
	// await act(async () => {
	// 	await axios.get.mockResolvedValue({
	// 		data: testRooms
	// 	});
	// 	render(
	// 		< Lobby />
	// 	)
	// });
	// console.log(component.debug())
	// expect(component.g).toBeInTheDocument()

	// const promise = Promise.resolve()
	// console.log(screen.debug())
	// expect(screen.queryByText(/No open room available/i)).not.toBeInTheDocument()
	// try {
	// 	await waitFor(() => {
	// 		expect(screen.queryByText(/No open room available/i)).not.toBeInTheDocument()

	// 	});
	// } catch (error) {
	// 	console.log(error)
	// }
	// });

})
