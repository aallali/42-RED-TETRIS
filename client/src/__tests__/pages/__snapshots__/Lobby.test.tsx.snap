// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_Loby PageXContainer test _ Loby Snapshot Check 1`] = `
<div
  className="  min-h-screen flex justify-center items-center shadow-xl"
>
  <div
    className="mt-0  flex flex-col lg:flex-row w-full justify-center gap-2 p-4"
  >
    <div
      className="bg-white lg:w-7/12 p-4 rounded-lg order-2 lg:order-first"
    >
      <h1
        className="text-gray-700 text-xl  font-bold"
      >
        Start your own room now !
      </h1>
      <div
        className="flex my-1  rounded-sm pt-3 pb-3 pr-4 pl-4  bg-blue-50 justify-between items-center"
      >
        <div
          className="flex flex-justify-start gap-1 items-center"
        >
          <div
            className="relative inline-block w-10 mr-1 align-middle select-none transition duration-200 ease-in"
          >
            <input
              className="toggle-checkbox absolute block w-6 h-6 rounded-full bg-white border-4 appearance-none cursor-pointer"
              data-testid="toggle"
              defaultChecked={true}
              id="toggle"
              name="toggle"
              onChange={[Function]}
              type="checkbox"
              value="multiplayer"
            />
            <label
              className="toggle-label block overflow-hidden h-6 rounded-full bg-gray-300 cursor-pointer"
              htmlFor="toggle"
            />
          </div>
          <span
            className="uppercase font-bold text-gray-700"
          >
            <label
              className=""
              htmlFor="toggle"
            >
              multiplayer
            </label>
          </span>
        </div>
        <div>
          <form
            className="m-0 flex"
          >
            <input
              className="text-xl w-full rounded-l-md p-2 border-t mr-0 border-b border-l text-gray-800 border-gray-200 bg-white"
              id="roomNameInput"
              onChange={[Function]}
              placeholder="room name..."
              value=""
            />
            <button
              className="text-md w-1/2 font-bold px-1 rounded-r-md bg-blue-400  text-gray-800 font-bold p-1 uppercase border-none"
              onClick={[Function]}
            >
              Create
            </button>
          </form>
          <p
            className="text-xs text-red-600"
          >
            
          </p>
        </div>
      </div>
      <h1
        className="text-gray-700 font-bold text-xl tracking-wider"
      >
        Rooms Online now :
      </h1>
      <div
        className="my-0"
      >
        <h1
          className="text-md"
        >
          No open room available ... create yours now!
        </h1>
      </div>
    </div>
    <div
      className="w-full lg:w-1/12 order-1 lg:order-last flex flex-col flex-1 justify-start gap-2"
    >
      <div
        className="bg-white p-1 rounded-md text-center"
      >
        <img
          alt="deflat"
          className="h-24  w-full object-cover content-center rounded-t-md"
          src="cover_profile.png"
        />
        <div
          className="flex justify-center"
        >
          <img
            alt="Mr.Abdellah Allali"
            className="w-14 h-14 rounded-full object-cover content-center -mt-8 border-4 border-white"
            src="https://ui-avatars.com/api/?background=B82E1F&bold=true&color=fff&name="
          />
        </div>
        <h1
          className="text-center font-bold tracking-wider text-gray-700   m-0"
        >
          
          
           
        </h1>
        <button
          className="bg-blue-100  border-1 border-blue-700 py-2 px-2 rounded-md text-blue-700 text-md font-bold mb-2"
        >
          Max Level: 
          0
        </button>
        <div
          className="w-full mx-auto"
        >
          <div
            className="flex flex-col flex-1 sm:flex-row gap-1"
          >
            <button
              className="bg-red-500 hover:bg-red-400 w-full py-1 px-4 mr-0 rounded-sm text-white text-sm font-semibold"
              onClick={[Function]}
            >
              Leave
            </button>
          </div>
        </div>
      </div>
      <div
        className="gap-2 justify-between flex flex-row"
      >
        <div
          className="w-3/5"
        >
          <div
            className="bg-white rounded-lg p-1 pl-0 pr-0 w-full "
          >
            <h1
              className="font-bold tracking-wider text-gray-800 text-md p-2"
            >
              Top  Ranked Players :
            </h1>
            <div
              className="flex flex-col items-center"
            >
              <div
                className="w-full md:w-full flex flex-col items-center h-60"
              >
                <div
                  className="w-full px-0"
                >
                  <div
                    className="flex flex-col items-center relative"
                  >
                    <table
                      className="table-fixed w-full"
                    >
                      <thead
                        className=""
                      >
                        <tr
                          className="border-b-0 border-black text-sm"
                        >
                          <th
                            className="w-2/12 bg-gray-100 text-left px-1"
                          >
                            Rank
                          </th>
                          <th
                            className="w-2/12 bg-gray-100 text-left px-1"
                          >
                            level
                          </th>
                          <th
                            className="w-5/12 bg-gray-100 text-left px-1"
                          >
                            nickname
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        className="divide-y divide-gray-400"
                      />
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="bg-white rounded-lg p-1 pl-0 pr-0 w-full "
        >
          <h1
            className="font-bold tracking-wider text-gray-800 text-xl p-2"
          >
            About :
          </h1>
          <div
            className="flex flex-col items-center"
          >
            <div
              className="w-full md:w-full flex flex-col items-center h-60"
            >
              <div
                className="w-full px-2 overflow-x-auto"
              >
                <p>
                  <b>
                    Red-Tetris
                  </b>
                   is an online multiplayer tetris game, realised in following stack [
                  <b>
                    reactJs
                  </b>
                  , 
                  <b>
                    redux
                  </b>
                  , 
                  <b>
                    sockets
                  </b>
                  , 
                  <b>
                    express
                  </b>
                  ] and written in typescript, it provides many features such playing in open/locked room and specify the size of your current room where you are the owner, play the game synchronously with joined members with same pieces, where one of you wins the game at the end, each time you clear a row, your opponents receive a n -1 row in their board, so be fast and clear more rows faster, the game just recently deployed to public so there is more feautures incomming to add and more bugs to fix :)
                </p>
                <p>
                  if you ever wanted to contact the creator, please feel free to reach me via : 
                  <b>
                    hi@allali.me
                  </b>
                   
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
